### ISSUE TITLE
Please Tell Us About Your Use of OpenQASM

### ISSUE THREAD

**Issue by jwoehr** [openqasm/openqasm#556]
### What is the expected behavior?

This is for users and implementers of OpenQASM to let the Technical Steering Committee know what you are doing with OpenQASM

The TSC is interested in:

- projects coded in OpenQASM
- implementations of OpenQASM
  - successful implementations
  - roadblocks you may have faced 
- tooling you have designed
- tooling you wish was present

... and any other ideas or comments you have based on active use of OpenQASM or implementation of OpenQASM parsers, compilers, tooling, etc.

**Comment by levbishop** [comment]
Yes, please! Do consider entering your use cases as comments on this issue - the TSC is very keen to make sure the language is meeting user needs and in getting a broader sense of how openQASM is solving real-world problems. The slack channel https://qiskit.enterprise.slack.com/archives/CG8JSE0UB is also available for more free-form discussion

**Comment by TheGupta2012** [comment]
### PyQASM
Hi there! We at **qBraid** have recently designed and developed a QASM3 semantic validation and unrolling tool called [`pyqasm`](https://github.com/qBraid/pyqasm). The motivation to develop it was the _lack of semantic analysis_ features in openqasm repo and an _incomplete feature set_ in the qiskit_qasm3_import library. Our current version depends directly on **openqasm3 v1.0.0** and [supports a large array of QASM3 constructs](https://github.com/qBraid/pyqasm/blob/main/pyqasm/README.md) where in a user can either - 
1. [validate](https://github.com/qBraid/pyqasm/tree/main/examples#program-validation) their program against QASM3 semantics 
2. [unroll](https://github.com/qBraid/pyqasm/tree/main/examples#program-unrolling) their generic QASM programs into a list of statements which only consist of qubit level instructions such as QuantumGate, QuantumBarrier, QuantumReset, etc. 

We are in the process of adding more features and support for other constructs of the language and our final goal is to develop a package that serves as a complete tool for processing, unrolling and semantic validation of QASM programs. 

If this tool can be used to enhance the capabilities of openqasm or Qiskit, we would love to explore that avenue with you guys!

**Comment by meamy** [comment]
I'll add my $.02 here.

We added support for openQASM 3 in [Feynman](https://github.com/meamy/feynman) as a way to explore [optimization of hybrid programs](https://arxiv.org/abs/2410.23493). Our tooling beyond the optimizations is minimal, but includes a gate inliner and loop unroller, plus a bit of control flow analysis.

Our experience was that the language was too large and cumbersome to easily experiment with these types of analyses. We're now debating whether it would be useful in the long-term to explore developing a reduced intermediate language or an analysis frontend (e.g. [CIL](https://github.com/cil-project/cil)), or to move to some other program model.

FWIW, my ideal use cases relate to circuit families, which as far as I understand is not being considered anymore.

**Comment by hodgestar** [comment]
@meamy Although the circuit families working group has ended it produced https://github.com/openqasm/openqasm/pull/346. The first part of this is being implemented in https://github.com/openqasm/openqasm/pull/563. If you are interested, you could let us know if what's being implemented would cover Feynman's needs.

OpenQASM 3 is a large language and the profile working group is looking at defining a useful subset to implement. If you have any thoughts on what a useful subset would look like, I'd love to hear your thoughts.

**Comment by meamy** [comment]
@hodgestar Thanks for the pointers! I've been keeping an eye on the progress of extended gate syntax, which almost gets to the point I was hoping for (templates on the register size).

My interests don't account for many (or in fact, most) of the use cases of openQASM 3, but if there's a working group looking at subsets to implement it may be relevant, so I'm happy to share my thoughts. In the end a large language isn't an issue provided decent tooling for program analysis, but I (and my group) only have so many cycles, and it may be too early for this kind of tooling anyway.

**Comment by jwoehr** [comment]
#559

**Comment by blakejohnson** [comment]
I just learned about a new project that supports OpenQASM 3: https://github.com/TeamGraphix/graphix-qasm-parser
